quic/s2n-quic-platform/src/io/turmoil.rs

```diff
@@ -136,117 +142,73 @@ impl Io {
-
-            // Poll for writablity if we have occupied slots available
-            let tx_interest = tx.occupied_len() > 0;
-            let tx_task = async {
-                if tx_interest {
-                    socket.writable().await
-                } else {
-                    futures::future::pending().await
-                }
-            };
-
-            let wakeups = endpoint.wakeups(&clock);
-            // pin the wakeups future so we don't have to move it into the Select future.
-            tokio::pin!(wakeups);
-
-            let timer_ready = timer.ready();
-
-            let select::Outcome {
-                rx_result,
-                tx_result,
-                timeout_expired,
-                application_wakeup,
-            } = if let Ok(res) = Select::new(rx_task, tx_task, &mut wakeups, timer_ready).await {
-                res
+/// Turmoil doesn't allow to split sockets for Tx and Rx so we need to spawn a single task to
+/// handle both jobs
+async fn run_io(
+    socket: UdpSocket,
+    mut producer: Producer<Message>,
+    mut consumer: Consumer<Message>,
+) -> io::Result<()> {
+    let mut poll_producer = false;
+
+    loop {
+        let socket_ready = socket.readable();
+        let consumer_ready = poll_fn(|cx| consumer.poll_acquire(1, cx));
+        let producer_ready = async {
+            if poll_producer {
+                poll_fn(|cx| producer.poll_acquire(1, cx)).await
             } else {
-                // The endpoint has shut down
-                return Ok(());
-            };
-
-            let wakeup_timestamp = clock.get_time();
-            let subscriber = endpoint.subscriber();
-            let mut publisher = event::EndpointPublisherSubscriber::new(
-                event::builder::EndpointMeta {
-                    endpoint_type: E::ENDPOINT_TYPE,
-                    timestamp: wakeup_timestamp,
-                },
-                None,
-                subscriber,
-            );
-
-            publisher.on_platform_event_loop_wakeup(event::builder::PlatformEventLoopWakeup {
-                timeout_expired,
-                rx_ready: rx_result.is_some(),
-                tx_ready: tx_result.is_some(),
-                application_wakeup,
-            });
-
-            if tx_result.is_some() {
-                tx.tx(&socket, &mut publisher)?;
+                core::future::pending().await
             }
+        };
 
-            if rx_result.is_some() {
-                rx.rx(&socket, &mut publisher)?;
-                endpoint.receive(&mut rx.rx_queue(), &clock);
+        let is_readable = Select::new(
+            consumer_ready,
+            producer_ready,
+            core::future::pending(),
+            socket_ready,
+        )
+        .await
+        .unwrap()
+        .timeout_expired;
+
+        if is_readable {
+            let mut count = 0;
+            for entry in producer.data() {
+                if let Ok((len, addr)) = socket.try_recv_from(entry.payload_mut()) {
```

Comment:
It looks like we expect the errors from `try_recv_from` and `try_send_to` to be recoverable. Could we add a comment explaining what those errors are??